File: android/src/mapkit/full/expo/modules/yandexmapkit/ExpoYandexSearchModule.kt
package expo.modules.yandexmapkit

import com.yandex.mapkit.GeoObject
import com.yandex.mapkit.geometry.BoundingBox
import com.yandex.mapkit.geometry.Geometry
import com.yandex.mapkit.geometry.Point
import com.yandex.mapkit.search.Response
import com.yandex.mapkit.search.SearchFactory
import com.yandex.mapkit.search.SearchManager
import com.yandex.mapkit.search.SearchManagerType
import com.yandex.mapkit.search.SearchOptions
import com.yandex.mapkit.search.SearchType
import com.yandex.mapkit.search.Session
import com.yandex.mapkit.search.ToponymObjectMetadata
import com.yandex.runtime.Error
import expo.modules.kotlin.Promise
import expo.modules.kotlin.functions.Queues
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

// Full-flavor implementation of the Search module (text search + reverse geocoding). Compiled only
// under the `full` MapKit flavor (android/build.gradle picks src/mapkit/full); the lite flavor gets
// the rejecting stub from src/mapkit/lite. See [[v2-full-flavor-architecture]] for the split.
class ExpoYandexSearchModule : Module() {
  private var searchManager: SearchManager? = null

  override fun definition() = ModuleDefinition {
    Name("ExpoYandexSearch")

    // Text search near a geometry (a point / bounding box, else the whole world).
    AsyncFunction("searchText") { query: String, options: SearchOptionsRecord?, promise: Promise ->
      manager().submit(query, resolveGeometry(options), searchOptions(options), listener(promise))
    }.runOnQueue(Queues.MAIN)

    // Reverse geocoding: the objects at a coordinate.
    AsyncFunction("searchPoint") { point: PointRecord, options: SearchOptionsRecord?, promise: Promise ->
      manager().submit(
        Point(point.latitude, point.longitude),
        options?.zoom?.toInt(),
        searchOptions(options),
        listener(promise)
      )
    }.runOnQueue(Queues.MAIN)

    // Resolve a MapKit object URI (e.g. a `ymapsbm1://…` from a suggest result) to full details.
    AsyncFunction("resolveURI") { uri: String, options: SearchOptionsRecord?, promise: Promise ->
      manager().resolveURI(uri, searchOptions(options), listener(promise))
    }.runOnQueue(Queues.MAIN)
  }

  // MapKitFactory (main module initialize()/build-time key) is the prerequisite; SearchFactory has
  // no initialize() of its own.
  private fun manager(): SearchManager =
    searchManager
      ?: SearchFactory.getInstance().createSearchManager(SearchManagerType.COMBINED)
        .also { searchManager = it }

  private fun listener(promise: Promise) = object : Session.SearchListener {
    override fun onSearchResponse(response: Response) {
      promise.resolve(serialize(response))
    }

    override fun onSearchError(error: Error) {
      promise.reject("E_SEARCH", "expo-yandex-mapkit: search failed (${error.javaClass.simpleName})", null)
    }
  }

  private fun resolveGeometry(options: SearchOptionsRecord?): Geometry {
    val box = options?.boundingBox
    val sw = box?.southWest
    val ne = box?.northEast
    if (sw != null && ne != null) {
      return Geometry.fromBoundingBox(
        BoundingBox(Point(sw.latitude, sw.longitude), Point(ne.latitude, ne.longitude))
      )
    }
    options?.userPosition?.let { return Geometry.fromPoint(Point(it.latitude, it.longitude)) }
    return Geometry.fromBoundingBox(BoundingBox(Point(-90.0, -180.0), Point(90.0, 180.0)))
  }

  private fun searchOptions(options: SearchOptionsRecord?): SearchOptions {
    val result = SearchOptions().setSearchTypes(resolveSearchTypes(options?.searchTypes))
    options?.userPosition?.let { result.setUserPosition(Point(it.latitude, it.longitude)) }
    options?.resultPageSize?.let { result.setResultPageSize(it.toInt()) }
    return result
  }

  private fun resolveSearchTypes(types: List<String>?): Int {
    val list = types ?: listOf("geo")
    var result = 0
    for (type in list) {
      result = result or when (type) {
        "geo" -> SearchType.GEO.value
        "biz" -> SearchType.BIZ.value
        else -> 0
      }
    }
    return if (result == 0) SearchType.GEO.value else result
  }

  private fun serialize(response: Response): List<Map<String, Any?>> =
    response.collection.children.mapNotNull { item -> item.obj?.let { serializeGeoObject(it) } }

  private fun serializeGeoObject(obj: GeoObject): Map<String, Any?> {
    val result = mutableMapOf<String, Any?>()
    obj.name?.let { result["name"] = it }
    obj.descriptionText?.let { result["description"] = it }
    val toponym = obj.metadataContainer.getItem(ToponymObjectMetadata::class.java)
    val point = toponym?.balloonPoint ?: obj.geometry.firstOrNull()?.point
    point?.let { result["point"] = mapOf("latitude" to it.latitude, "longitude" to it.longitude) }
    toponym?.address?.let { address ->
      result["formattedAddress"] = address.formattedAddress
      result["addressComponents"] = address.components.map { component ->
        mapOf(
          "name" to component.name,
          // Kind enum name lowercased → snake_case (e.g. METRO_STATION → "metro_station"), matching iOS.
          "kinds" to component.kinds.map { kind -> kind.name.lowercase() }
        )
      }
    }
    return result
  }
}

File: ios/ExpoYandexSearchModule.swift
import ExpoModulesCore
#if YANDEX_MAPS_FULL
  import YandexMapsMobile
#endif

// Search option record (shared shape; no MapKit types, always compiled). Reuses PointRecord and the
// SuggestBoundingBoxRecord {southWest, northEast} shape.
internal struct SearchOptionsRecord: Record {
  @Field var userPosition: PointRecord?
  @Field var boundingBox: SuggestBoundingBoxRecord?
  @Field var searchTypes: [String]?
  @Field var resultPageSize: Int?
  @Field var zoom: Int?
}

// The Search module (text search + reverse geocoding). Real work is behind `#if YANDEX_MAPS_FULL`
// (references classes the lite pod omits); in lite every call rejects clearly. See the map view
// module for the shared record types.
public class ExpoYandexSearchModule: Module {
  #if YANDEX_MAPS_FULL
    private var searchManager: YMKSearchManager?
    // Sessions must be retained until their response arrives — dropping one cancels the request.
    private var sessions: [YMKSearchSession] = []
  #endif

  public func definition() -> ModuleDefinition {
    Name("ExpoYandexSearch")

    AsyncFunction("searchText") { (query: String, options: SearchOptionsRecord?, promise: Promise) in
      #if YANDEX_MAPS_FULL
        self.retain(
          self.manager().submit(
            withText: query,
            geometry: self.resolveGeometry(options),
            searchOptions: self.searchOptions(options)
          ) { response, error in self.handle(response, error, promise) })
      #else
        promise.reject("E_FULL_REQUIRED", Self.fullRequiredMessage)
      #endif
    }.runOnQueue(.main)

    AsyncFunction("searchPoint") { (point: PointRecord, options: SearchOptionsRecord?, promise: Promise) in
      #if YANDEX_MAPS_FULL
        self.retain(
          self.manager().submit(
            with: YMKPoint(latitude: point.latitude, longitude: point.longitude),
            zoom: options?.zoom.map { NSNumber(value: $0) },
            searchOptions: self.searchOptions(options)
          ) { response, error in self.handle(response, error, promise) })
      #else
        promise.reject("E_FULL_REQUIRED", Self.fullRequiredMessage)
      #endif
    }.runOnQueue(.main)

    AsyncFunction("resolveURI") { (uri: String, options: SearchOptionsRecord?, promise: Promise) in
      #if YANDEX_MAPS_FULL
        self.retain(
          self.manager().resolveURI(
            withUri: uri, searchOptions: self.searchOptions(options)
          ) { response, error in self.handle(response, error, promise) })
      #else
        promise.reject("E_FULL_REQUIRED", Self.fullRequiredMessage)
      #endif
    }.runOnQueue(.main)
  }

  private static let fullRequiredMessage =
    "expo-yandex-mapkit: search requires the MapKit full flavor. "
    + "Set flavor: 'full' in the config plugin (and rebuild)."

  #if YANDEX_MAPS_FULL
    private func manager() -> YMKSearchManager {
      let manager = searchManager ?? YMKSearchFactory.instance().createSearchManager(with: .combined)
      searchManager = manager
      return manager
    }

    private func retain(_ session: YMKSearchSession) {
      sessions.append(session)
    }

    private func handle(_ response: YMKSearchResponse?, _ error: Error?, _ promise: Promise) {
      if let error = error {
        promise.reject(
          "E_SEARCH", "expo-yandex-mapkit: search failed — \(error.localizedDescription)")
        return
      }
      let items = (response?.collection.children ?? []).compactMap { $0.obj }.map { self.serialize($0) }
      promise.resolve(items)
    }

    private func resolveGeometry(_ options: SearchOptionsRecord?) -> YMKGeometry {
      if let sw = options?.boundingBox?.southWest, let ne = options?.boundingBox?.northEast {
        return YMKGeometry(
          boundingBox: YMKBoundingBox(
            southWest: YMKPoint(latitude: sw.latitude, longitude: sw.longitude),
            northEast: YMKPoint(latitude: ne.latitude, longitude: ne.longitude)))
      }
      if let up = options?.userPosition {
        return YMKGeometry(point: YMKPoint(latitude: up.latitude, longitude: up.longitude))
      }
      return YMKGeometry(
        boundingBox: YMKBoundingBox(
          southWest: YMKPoint(latitude: -90, longitude: -180),
          northEast: YMKPoint(latitude: 90, longitude: 180)))
    }

    private func searchOptions(_ options: SearchOptionsRecord?) -> YMKSearchOptions {
      let result = YMKSearchOptions()
      result.searchTypes = resolveSearchTypes(options?.searchTypes)
      if let up = options?.userPosition {
        result.userPosition = YMKPoint(latitude: up.latitude, longitude: up.longitude)
      }
      if let size = options?.resultPageSize {
        result.resultPageSize = NSNumber(value: size)
      }
      return result
    }

    private func resolveSearchTypes(_ types: [String]?) -> YMKSearchType {
      let list = types ?? ["geo"]
      var result: YMKSearchType = []
      for type in list {
        switch type {
        case "geo": result.insert(.geo)
        case "biz": result.insert(.biz)
        default: break
        }
      }
      return result.isEmpty ? [.geo] : result
    }

    private func serialize(_ obj: YMKGeoObject) -> [String: Any?] {
      var result: [String: Any?] = [:]
      if let name = obj.name {
        result["name"] = name
      }
      if let description = obj.descriptionText {
        result["description"] = description
      }
      let toponym =
        obj.metadataContainer.getItemOf(YMKSearchToponymObjectMetadata.self)
        as? YMKSearchToponymObjectMetadata
      if let point = toponym?.balloonPoint ?? obj.geometry.first?.point {
        result["point"] = ["latitude": point.latitude, "longitude": point.longitude]
      }
      if let address = toponym?.address {
        result["formattedAddress"] = address.formattedAddress
        result["addressComponents"] = address.components.map { component -> [String: Any?] in
          ["name": component.name, "kinds": component.kinds.map { self.kindName($0) }]
        }
      }
      return result
    }

    // iOS boxes each address-component kind as an NSNumber (the YMKSearchComponentKind raw value).
    // Map it to the same snake_case string Android produces from the enum's `name`.
    private func kindName(_ boxed: NSNumber) -> String {
      guard let kind = YMKSearchComponentKind(rawValue: UInt(truncating: boxed)) else {
        return "unknown"
      }
      switch kind {
      case .unknown: return "unknown"
      case .country: return "country"
      case .region: return "region"
      case .province: return "province"
      case .area: return "area"
      case .locality: return "locality"
      case .district: return "district"
      case .street: return "street"
      case .house: return "house"
      case .entrance: return "entrance"
      case .level: return "level"
      case .apartment: return "apartment"
      case .route: return "route"
      case .station: return "station"
      case .metroStation: return "metro_station"
      case .railwayStation: return "railway_station"
      case .vegetation: return "vegetation"
      case .hydro: return "hydro"
      case .airport: return "airport"
      case .other: return "other"
      @unknown default: return "unknown"
      }
    }
  #endif
}
